### ISSUE TITLE
Unify activities and exercises?

### ISSUE THREAD

**Issue by tischi** [NEUBIAS/training-resources#325]
Some observations:

1. We felt we also want exercise prefaces (as for activities) to make them more uniform across platforms 
2. We sometimes have multiple activities and teach only some of them
3. We also sometimes have multiple exercises and may ask students to only do some of them
4. Personally, I find it annoying that we have different folders and slightly different structures (w and w/o solutions) for the activities and exercises and for new contributors to our material getting and remembering the difference is a bit of a hurdle.

I wonder: could we simplify it to do it like this?

- Activitity 1: sfsdg
  - Explanation of the activity/exercise in general terms (maybe even including a figure)
  - Show activity for platform: dropdown
- Activitity 2: sfsfgd
  - Explanation of the activity/exercise  n general terms (maybe even including a figure)
  - Show activity for platform: dropdown
- Activitity 3: sdfsgsg
  - Explanation of the activity/exercise in general terms (maybe even including a figure)
  - Show activity for platform: dropdown
 
And then, when teaching, one could choose one or two activities to click along and then tell the students to do some of the activities on their own (and one could say: if they want to challenge themselves they can try to to the activity just following the general explanation, or they look into the "solution", which now would simply be the drop-down. In fact, whatever is contained in the dropdown may still have a solution section, if approriate.

@manerotoni @AnniekStok @maulakhan what do you think?

**Comment by AnniekStok** [comment]
I like the idea! It would be more simple to not distinguish activity/exercise separately, and this way they can still easily be used differently depending on the situation.

**Comment by manerotoni** [comment]
I like the idea too. Yes the difference between activity and exercise is often artificial.
This is also a more in person/online discussion. We can collect some ideas now and then do 
a meeting after we finish the IJ-macro course.

I think a good example are the workflows. 

I think having the possibility of including some figures in the activity-preface would be great. For instance you may want to show: this is your starting image, find a way by using this and that method to reach following result. 

In the dropdown the layout may slightly change. For instance a more detailed description of the activity with hints where to click, with a solution/Code. I would not force it in a specific way. It could just be plain code, ..

**Comment by tischi** [comment]
I would love to also hear @miura opinion about this. 
Please have a look at this module to get an idea how it currently looks like: https://neubias.github.io/training-resources/binarization/index.html

One thing we have to keep in mind with the new proposal (see above): if we decide to change an activity/exercise we will have to change it for all implementations! That can become quite a task/burden, but on the upside we will have a very nice and consistent material (personally, I find it _super interesting_ to quickly see how the same task is done in different platforms/languages, so for me it'd be totally worth it).

**Comment by tischi** [comment]
I managed to set it up that it looks like this (obviously normally they would be different and there can be as many as we wish).

<img width="1068" alt="image" src="https://user-images.githubusercontent.com/2157566/157541524-ccb5ff2c-c878-4f36-9929-aeeebfbf9076.png">

The syntax within the document is like this (a list of: one preface document followed by several implementations):

```
  - ["watershed/exercises/shape_preface.md",[["ImageJ Macro","watershed/exercises/morpholibj_shape_watershed_exercise.ijm"],["ImageJ Jython","watershed/exercises/morpholibj_shape_watershed_exercise.py"]]]
  - ["watershed/exercises/shape_preface.md",[["ImageJ Macro", "watershed/exercises/morpholibj_shape_watershed_exercise.ijm"],["ImageJ Jython", "watershed/exercises/morpholibj_shape_watershed_exercise.py"]]]
```

What do you think? I think that's already more convenient than what we have.
Little downside: also the prefaces need to go into a separate documents now.

**Comment by manerotoni** [comment]
Hi @tischi,
I think it is ok to have several files associated with activity/file. Can you add an image also to the preface? This should be possible as normal MD.
If one decides to modify an exercise, by changing a figure, I opt to create an additional exercise. This is easier than changing again and again the same exercise.

**Comment by tischi** [comment]
@manerotoni 
I saw that you added **```java** to the *.ijm exercise files in the watershed module. While this is nice for rendering one cannot execute those files anymore in ImageJ. Instead I tried to improve the code where Jekyll renders the modules to automatically apply the correct rendering. OK for you?

```
202 {% for platform in page.exercises %}$
203 <blockquote class="challenge" id="{{ platform[0] | slugify }}-exercises-div">$
204 <h2>{{ platform[0] }}</h2>$
205 {% assign includefile = platform[1] %}$
206 {% capture exercise %}{% include {{ includefile }} %}{% endcapture %}$
207 {% assign ext = includefile | split: "." | last %}$
208 {% if ext == "ijm" %}$
209 {% highlight java %}$
210 {{ exercise }}$
211 {% endhighlight %}$
212 {% elsif ext == "py" %}$
213 {% highlight python %}$
214 {{ exercise }}$
215 {% endhighlight %}$
216 {% elsif ext == "md" %}$
217 {{ exercise | markdownify }}$
218 {% else %}$
219 {{ exercise }}$
220 {% endif %}
```

**Comment by manerotoni** [comment]
Yes I realized it yesterday. Great job

**Comment by tischi** [comment]
@AnniekStok 

The name of the branch where I started working on this is: `new_module_layout`

Only the watershed module currently is rendered in this branch.

The file to work on is: `_layouts/module.html` in conjunction with `_modules/watershed.md`.

I started making the changes for the exercises (not for the activities). We would then delete the activities from `_layouts/module.html`. I had a look and it currently does not fully work yet.

-----

@manerotoni @stemarcotti 

I am having some concerns: I think it is nice if there is exactly one activity and one exercise, because this makes it easy for a teacher to teach a module. Having a lot of choices puts extra burden on the teacher. 

Maybe we should revisit the modules where we have more than one activity and check whether this is too much content for one module?

Maybe: multiple exercises is OK, but we should have only one activity?

Even if we unify the code for activities and exercises, which I still think could be a good idea, I think it would be nice to somehow indicate what "we" recommend to done by the teacher and what we think to be done as exercises by the students.

Unless, we want the material also to be used for self-study, then the difference between activities and exercises is gone anyway, is it? 

Don't know, bit confused 😕
